### PR TITLE
enabled quantile regression in gbm

### PR DESCRIPTION
**Pull request by bthieurmel**
We are always happy to receive pull requests.

Please make sure you have read our coding guidelines: 
https://github.com/mlr-org/mlr/wiki/mlr-Coding-Guidelines

This especially means that you have understood:

* The style guide - our lintr will provide you feedback on this
* How to run tests locally. Yes, travis will also run them for you, bad it is annoying to wait for this.
* How to run R CMD CHECK locally. See point before. 

Also it's helpful to get into direct contact with the suggested reviewers to get help, getting your PR merged.
You might want to join our slack at: 
https://mlr-org.slack.com

### CODE REVIEW

File: R/RLearner_regr_gbm.R [hunk 1]
```diff
@@ -13,14 +13,16 @@ makeRLearner.regr.gbm = function() {
       makeNumericLearnerParam(id = "shrinkage", default = 0.001, lower = 0),
       makeNumericLearnerParam(id = "bag.fraction", default = 0.5, lower = 0, upper = 1),
       makeNumericLearnerParam(id = "train.fraction", default = 1, lower = 0, upper = 1),
+      makeNumericLearnerParam(id = "alpha", default = 0.5, lower = 0, upper = 1),
```
Comment (pat-s) [inline R/RLearner_regr_gbm.R]: Please make this param dependent on `quantile`. No need to add a statement in the learner note.

File: R/RLearner_regr_gbm.R [hunk 2]
```diff
@@ -30,10 +32,39 @@ trainLearner.regr.gbm = function(.learner, .task, .subset, .weights = NULL, ...)
...
+    } else {
+      alpha <- 0.5
+    }
+    if(params$distribution %in% "quantile"){
+      params$distribution <- list(name = "quantile", alpha = alpha)
+    }
+    params$formula <- f
+    params$data <- getTaskData(.task, .subset)
+
+    do.call(gbm::gbm, params)
```
Comment (pat-s) [inline R/RLearner_regr_gbm.R]: Why is this extra code needed here?

File: R/RLearner_regr_gbm.R [hunk 3]
```diff
@@ -30,10 +32,39 @@ trainLearner.regr.gbm = function(.learner, .task, .subset, .weights = NULL, ...)
   f = getTaskFormula(.task)
   if (is.null(.weights)) {
     f = getTaskFormula(.task)
-    gbm::gbm(f, data = getTaskData(.task, .subset), ...)
+
+    params <- list(...)
+    if("alpha" %in% names(params)){
+      alpha <- params$alpha
```
Comment (larskotthoff) [inline R/RLearner_regr_gbm.R]: Please check the style guide -- we use = for assignment.

File: R/RLearner_regr_gbm.R [hunk 4]
```diff
@@ -30,10 +32,39 @@ trainLearner.regr.gbm = function(.learner, .task, .subset, .weights = NULL, ...)
   f = getTaskFormula(.task)
   if (is.null(.weights)) {
     f = getTaskFormula(.task)
-    gbm::gbm(f, data = getTaskData(.task, .subset), ...)
+
+    params <- list(...)
+    if("alpha" %in% names(params)){
```
Comment (larskotthoff) [inline R/RLearner_regr_gbm.R]: Space after if.

File: R/RLearner_regr_gbm.R [hunk 5]
```diff
@@ -30,10 +32,39 @@ trainLearner.regr.gbm = function(.learner, .task, .subset, .weights = NULL, ...)
...
+    if(params$distribution %in% "quantile"){
+      params$distribution <- list(name = "quantile", alpha = alpha)
+    }
+    params$formula <- f
+    params$data <- getTaskData(.task, .subset)
+    params$weights <- .weights
+
+    do.call(gbm::gbm, params)
   }
 }
```
Comment (larskotthoff) [inline R/RLearner_regr_gbm.R]: Can you also please change the test to test this new parameter value?

File: R/RLearner_regr_gbm.R (hunk 2 above)
Comment (bthieurmel) [reply R/RLearner_regr_gbm.R]: ``gbm`` use two ways to pass ``distribution`` parameters : 

- simple character as "gaussian" or "laplace"
- named list, for "quantile" like ``list(name = "quantile", alpha = 0.5)

So we have to control the distribution and format well ``distrbution`` argument

File: R/RLearner_regr_gbm.R (hunk 4 above)
Comment (bthieurmel) [reply R/RLearner_regr_gbm.R]: Done. Perhaps you can update the link on the PR message ? 

````
Please make sure you have read our coding guidelines:
https://github.com/mlr-org/mlr/wiki/mlr-Coding-Guidelines
````

and remplace with https://www.notion.so/Style-Guide-740bc663207a4bbb9a457987bda6fd91

File: R/RLearner_regr_gbm.R (hunk 5 above)
Comment (bthieurmel) [reply R/RLearner_regr_gbm.R]: Done

File: R/RLearner_regr_gbm.R (hunk 4 above)
Comment (pat-s) [reply R/RLearner_regr_gbm.R]: Thanks, done.

File: R/RLearner_regr_gbm.R (hunk 2 above)
Comment (pat-s) [reply R/RLearner_regr_gbm.R]: ok.

File: NEWS.md [hunk 6]
```diff
@@ -11,6 +11,7 @@
   See `?regr.randomForest` for more details.  
   `regr.ranger` relies on the functions provided by the package ("jackknife" and "infjackknife" (default))  
   (@jakob-r, #1784)
+- `regr.gbm` now supports `quantile distribution`
```
Comment (pat-s) [inline NEWS.md]: Also add your username and the PR number please.

**Comment by pat-s** [comment]
Thanks. Only one comment left from my side.

Please take a look why the test is failing and tell us if you need help. 

Looks almost ready to merge :+1:

**Comment by bthieurmel** [comment]
Seems to be ok (missing ``!`` and so don't pass weights information....)

File: R/RLearner_regr_gbm.R [hunk 7]
```diff
@@ -28,13 +30,25 @@ makeRLearner.regr.gbm = function() {
 #' @export
 trainLearner.regr.gbm = function(.learner, .task, .subset, .weights = NULL, ...) {
   f = getTaskFormula(.task)
-  if (is.null(.weights)) {
-    f = getTaskFormula(.task)
-    gbm::gbm(f, data = getTaskData(.task, .subset), ...)
+
+  params = list(...)
+  if("alpha" %in% names(params)) {
```
Comment (larskotthoff) [inline R/RLearner_regr_gbm.R]: Space after if and I would check whether `params$alpha` is `NULL` for consistency with how it's set below.

File: R/RLearner_regr_gbm.R [hunk 8]
```diff
@@ -28,13 +30,25 @@ makeRLearner.regr.gbm = function() {
...
+    alpha = params$alpha
+    params$alpha = NULL
   } else {
-    f = getTaskFormula(.task)
-    gbm::gbm(f, data = getTaskData(.task, .subset), weights = .weights, ...)
+    alpha = 0.5
+  }
+  if(params$distribution %in% "quantile") {
```
Comment (larskotthoff) [inline R/RLearner_regr_gbm.R]: Space after if and should that check be simply `==`?

**Comment by pat-s** [comment]
For style, you can simply apply the [mlr style guide](https://www.notion.so/mlrorg/Style-Guide-740bc663207a4bbb9a457987bda6fd91). (in case you haven't already)

**Comment by pat-s** [comment]
@larskotthoff 	If you approve, feel free to merge. We can run the style fixer afterwards. Applying it on a forked branch is not easy.

**Comment by larskotthoff** [comment]
Merging.

**Pull request merged**